stablehlo/reference/Ops.cpp

```diff
@@ -1420,6 +1475,62 @@ Tensor evalSelectOp(const Tensor &pred, const Tensor &onTrue,
   return result;
 }
 
+Tensor evalSelectAndScatter(const Tensor &operand, const Tensor &source,
+                            const Tensor &initValue,
+                            const Sizes &windowDimensions,
+                            const Sizes &windowStrides, const Sizes &paddingLow,
+                            const Sizes &paddingHigh, Region &select,
+                            Region &scatter, Scope &scope,
+                            ShapedType resultType) {
+  Tensor result(resultType, initValue.get({}));
+  auto rank = operand.getRank();
+
+  for (auto sourceIt = IndexSpaceIterator(source.getShape(), Sizes(rank, 0));
+       sourceIt != IndexSpaceIterator(source.getShape(), std::nullopt);
+       ++sourceIt) {
+    std::optional<Element> selectedVal;
+    std::optional<Index> selectedIndex;
+    iterateThroughWindow(
+        windowDimensions, windowStrides, operand.getShape(), paddingLow,
+        *sourceIt, [&](const Index &operandIndex) {
+          auto currVal = operand.get(operandIndex);
+          if (!selectedVal.has_value()) {
+            selectedVal = currVal;
+            selectedIndex = operandIndex;
+          }
+
+          Tensor selectedValTensor(
+              RankedTensorType::get({}, selectedVal.value().getType()),
+              selectedVal.value());
+          Tensor currValTensor(RankedTensorType::get({}, currVal.getType()),
+                               currVal);
+          auto selectedResult =
+              eval(select, {selectedValTensor, currValTensor}, &scope);
+
+          bool selected = !selectedResult[0].get({}).getBooleanValue();
+          if (selected) {
+            selectedVal = currVal;
+            selectedIndex = operandIndex;
+          }
+        });
+    iterateThroughWindow(
+        windowDimensions, windowStrides, operand.getShape(), paddingLow,
+        *sourceIt, [&](const Index &operandIndex) {
+          if (operandIndex == selectedIndex) {
+            Tensor sourceValues(
+                RankedTensorType::get({2}, initValue.getElementType()));
+            sourceValues.set({0}, source.get(*sourceIt));
+            sourceValues.set({1}, result.get(operandIndex));
+
+            auto reducedResult =
+                evalReduceOp({sourceValues}, {initValue}, {0}, scatter, scope);
+            result.set(operandIndex, reducedResult[0].get({}));
+          }
+        });
+  }
+  return result;
+}
```

Comment:
Left feedback with simplifications and some style comments via a pull request: https://github.com/ghpvnist/stablehlo/pull/7.

Reply:
Thanks for the suggestions! Added some minor adjustments.